### PR TITLE
Make helm chart job only run if container is built ok

### LINKED ISSUE #136: CI: Helm chart should not be created if the container creation fails
Currently there seems to be no link between the two actions

update: description - from @callumforrester 

Typically in a Github release pipeline, you want actions to depend on one another so that broken artefacts are not published, for example there is no point in uploading to PyPi or publishing a container image if the tests won't even pass.
Currently we have a Github actions job for publishing the helm chart on a release, and it doesn't seem to have any dependencies because it will try to package and publish a chart even if other jobs fail. It should be one of the last steps in the pipeline and should depend on the container build at the very least.

## Acceptance Criteria

-     Helm publishing job depends on container successfully building and publishing
-     If the container fails, no new helm chart is published

### PR DESCRIPTION
**Pull request by stan-dot**
Fixes #136

### CODE REVIEW

**Comment by stan-dot** [comment]
that is the culprit - might need to solve both issues in one go for this to pass
https://github.com/DiamondLightSource/blueapi/issues/411

**Comment by callumforrester** [comment]
Please give this PR a sensible title

File: .github/workflows/_container.yml [hunk 1]
```diff
@@ -11,7 +18,16 @@ jobs:
...
+      - name: Validate PEP 440 version compliance
+        run: |
+          ref="${{ github.ref_name }}"
+          my_regex=${{env.PEP_440_REGEX}}
+          if [[ "$ref" =~ $my_regex ]]; then
+            echo "PEP 440 compliant version: $ref"
+          else
+            echo "Invalid PEP 440 version: $ref"
+            exit 1
+          fi
```
Comment (DiamondJoseph) [inline .github/workflows/_container.yml:30]: These changes seem unrelated, why are they required?

Comment (stan-dot) [reply .github/workflows/_container.yml:30]: otherwise it rans into #411 , so this PR fixes both,

Comment (callumforrester) [reply .github/workflows/_container.yml:30]: This doesn't fix #411 because if the user makes a release with a wrongly-formatted tag, the release will still appear in [Github](https://github.com/DiamondLightSource/blueapi/releases) and PyPi(https://pypi.org/project/blueapi/#history). I don't think there's any harm in checking here, although it would be redundant if we could do something like [this](https://github.com/DiamondLightSource/blueapi/issues/411#issuecomment-2429132946).

However, PEP440 actually allows for a wider range of error messages than helm (for instance, it allows `1.1a1` where helm would require `1.1-a1`) so this check and error message aren't quite right. See https://helm.sh/docs/chart_best_practices/conventions/

Comment (stan-dot) [reply .github/workflows/_container.yml:30]: well we convert to semver later (      - name: Convert PEP 440 version to SemVer) , is that ok?

Comment (callumforrester) [reply .github/workflows/_container.yml:30]: I would be okay with remove the PEP440 check entirely and making a separate issue for adding SemVer check. I would not be okay with leaving the PEP440 check in since it won't catch any useful errors reported in #411.

Comment (stan-dot) [reply .github/workflows/_container.yml:30]: why not semver check in this PR?

Comment (callumforrester) [reply .github/workflows/_container.yml:30]: That would also be acceptable. It's only the PEP 440 check that I won't approve.

**Review by callumforrester — CHANGES_REQUESTED** [review]

File: .github/workflows/_container.yml [hunk 2]
```diff
@@ -11,7 +18,16 @@ jobs:
...
+      - name: Validate SemVer2 version compliance
+        run: |
+          ref="${{ github.ref_name }}"
+          my_regex="${{env.SEMVER_REGEX}}"
+          if [[ "$ref" =~ $my_regex ]]; then
+            echo "SemVer compliant version: $ref"
+          else
+            echo "Invalid SemVer version: $ref"
+            exit 1
+          fi
```
[the remark this answers is not in the document]
Comment (stan-dot) [reply .github/workflows/_container.yml:30]: what's your act setup? I think I tried to run this once on my workstation and it failed

[the remark this answers is not in the document]
Comment (stan-dot) [reply .github/workflows/_container.yml:30]: > your regex

it's not _mine_ , it's literally an official suggested from `semver.org`.

and this script fails with the lower (official) regex on.

```bash

#!/bin/bash

# Define sample JSON data
json_data='{
  "ref": "refs/tags/2.3.4-a2",
  "ref_name": "2.3.4-a2"
}'

# Extract the "ref_name" value
ref=$(echo "$json_data" | jq -r '.ref_name')

# Define the SemVer regex pattern (ensuring to escape it correctly)
# SEMVER_REGEX="^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(-[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*)?(\+[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*)?$"
SEMVER_REGEX="^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"

# Test if the extracted ref_name matches the SemVer pattern
if [[ "$ref" =~ $SEMVER_REGEX ]]; then
  echo "SemVer compliant version: $ref"
else
  echo "Invalid SemVer version: $ref"
  exit 1
fi


```

[the remark this answers is not in the document]
Comment (stan-dot) [reply .github/workflows/_container.yml:30]: if I run it with the first regex instead (the one commented out), it works

[the remark this answers is not in the document]
Comment (stan-dot) [reply .github/workflows/_container.yml:30]: I presume that we'd like to skip the check on regular branches though

File: .github/workflows/_container.yml [hunk 3]
```diff
@@ -54,3 +70,36 @@ jobs:
...
+      - name: checkout repo
+        uses: actions/checkout@v3
+
+      - name: install helm
+        uses: Azure/setup-helm@v3
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+        id: install
+
+      - name: login to acr using helm
```
[the remark this answers is not in the document]
Comment (stan-dot) [reply .github/workflows/_container.yml]: added

File: .github/workflows/_container.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (stan-dot) [reply .github/workflows/_container.yml:31]: solved

[the remark this answers is not in the document]
Comment (callumforrester) [reply .github/workflows/_container.yml:31]: Works in act for me!

**Review by callumforrester — APPROVED** [review]
LGTM!
